### PR TITLE
BUG: replace with value also being replaced

### LINKED ISSUE #46306: BUG: replace method with np.nan produces incorrect results
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [ ] I have confirmed this bug exists on the main branch of pandas.

### Reproducible Example

```python
import pandas as pd
import numpy as np

df = pd.DataFrame({'A': [0, 1, 2], 'B': [1, 0, 2]})
df.replace({0:1, 1:np.nan})
```

### Issue Description

Pandas 1.4 changed the way the replace method works:

pandas 1.3.4:
`df`
    A  B
0  0  1
1  1  0
2  2  2
`df.replace({0:1, 1:np.nan})`
     A    B
0  1.0  NaN
1  NaN  1.0
2  2.0  2.0

pandas 1.4.1:
`df`
   A  B
0  0  1
1  1  0
2  2  2
`df.replace({0:1, 1:np.nan})`
     A    B
0  NaN  NaN
1  NaN  NaN
2  2.0  2.0	

As you can see with the latest version first value 0 was replaced to 1 and then existing '1' and the new from replaced 0 were replaced to nan.
However, the same replacement using other values then np.nan works ok, for instance:
`df.replace({0:1, 1:10})`
    A   B
0   1  10
1  10   1
2   2   2

### Expected Behavior

It should work the same way as in version 1.3.

### Installed Versions

<details>

INSTALLED VERSIONS
------------------
commit           : bb1f651536508cdfef8550f93ace7849b00046ee
python           : 3.9.1.final.0
python-bits      : 64
OS               : Windows
OS-release       : 10
Version          : 10.0.14393
machine          : AMD64
processor        : Intel64 Family 6 Model 79 Stepping 0, GenuineIntel
byteorder        : little
LC_ALL           : None
LANG             : None
LOCALE           : English_United States.1252

pandas           : 1.4.0
numpy            : 1.22.2
pytz             : 2021.3
dateutil         : 2.8.2
pip              : 20.2.3
setuptools       : 49.2.1
Cython           : None
pytest           : 7.0.0
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : None
IPython          : None
pandas_datareader: None
bs4              : None
bottleneck       : None
fastparquet      : None
fsspec           : None
gcsfs            : None
matplotlib       : None
numba            : None
numexpr          : None
odfpy            : None
openpyxl         : 3.0.9
pandas_gbq       : None
pyarrow          : None
pyreadstat       : None
pyxlsb           : None
s3fs             : None
scipy            : None
sqlalchemy       : None
tables           : None
tabulate         : None
xarray           : None
xlrd             : 2.0.1
xlwt             : None
zstandard        : None

</details>

### PR DESCRIPTION
**Pull request by lukemanley**
- [x] closes #46306
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added an entry in `doc/source/whatsnew/v1.4.2.rst`

### CODE REVIEW

**Review by jreback — CHANGES_REQUESTED** [review]
lgtm, can you add a whatsnew note in regression section of 1.4.2 and ping on green.

File: pandas/core/internals/blocks.py [hunk 1]
```diff
@@ -612,9 +612,15 @@ def replace(
 
         else:
             # split so that we only upcast where necessary
-            return self.split_and_operate(
-                type(self).replace, to_replace, value, inplace=True
-            )
+            blocks = []
+            nbs = self._split()
```
Comment (jreback) [inline pandas/core/internals/blocks.py:616]: could write this entire section as a list-comprehension, but maybe not as readable

Comment (lukemanley) [reply pandas/core/internals/blocks.py]: good idea, i like the list comprehension better

Comment (lukemanley) [reply pandas/core/internals/blocks.py]: Went back to a non-list comprehension implementation as it requires unpacking.

**Comment by lukemanley** [comment]
@jreback - greenish

**Comment by jreback** [comment]
@meeseeksdev backport 1.4.x

File: doc/source/whatsnew/v1.4.2.rst [hunk 2]
```diff
@@ -18,6 +18,7 @@ Fixed regressions
 - Fixed regression in :func:`read_csv` killing python process when invalid file input was given for ``engine="c"`` (:issue:`45957`)
 - Fixed memory performance regression in :meth:`Series.fillna` when called on a :class:`DataFrame` column with ``inplace=True`` (:issue:`46149`)
 - Provided an alternative solution for passing custom Excel formats in :meth:`.Styler.to_excel`, which was a regression based on stricter CSS validation. Examples available in the documentation for :meth:`.Styler.format` (:issue:`46152`)
+- Fixed regression in :meth:`DataFrame.replace` when a replacement value was also a target for replacement (:issue:`46335`)
```
Comment (simonjayhawkins) [inline doc/source/whatsnew/v1.4.2.rst:21]: will change the issue number to #46306 in a release note tidy up

Comment (simonjayhawkins) [reply doc/source/whatsnew/v1.4.2.rst:21]: done in #46354

**Pull request merged**